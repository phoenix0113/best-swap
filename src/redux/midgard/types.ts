import { RemoteData } from '@devexperts/remote-data-ts';
import BigNumber from 'bignumber.js';
import { Maybe } from '../../types/bepswap';
import {
  AssetDetail,
  PoolDetail,
  StakersAssetData,
  ThorchainEndpoints,
  ThorchainEndpoint,
  InlineResponse200,
  TxDetailsTypeEnum,
} from '../../types/generated/midgard';

export type AssetDetailMap = {
  [asset: string]: AssetDetail;
};

export type PoolDataMap = {
  [symbol: string]: PoolDetail;
};

export type GetStakerPoolDataPayload = {
  asset: string;
  address: string;
};

export type StakerPoolData = {
  [symbol: string]: StakersAssetData;
};

export type PriceDataIndex = {
  [symbol: string]: BigNumber;
};

export type TxDetailType = TxDetailsTypeEnum.Swap
 | TxDetailsTypeEnum.Stake | TxDetailsTypeEnum.Unstake
 | TxDetailsTypeEnum.Add | TxDetailsTypeEnum.Refund;

export type GetTxByAddressPayload = {
  address: string;
  offset: number;
  limit: number;
  type?: TxDetailType;
};

export type GetTxByAddressTxIdPayload = {
  address: string;
  txId: string;
  offset: number;
  limit: number;
  type?: TxDetailType;
};

export type GetTxByAddressAssetPayload = {
  address: string;
  asset: string;
  offset: number;
  limit: number;
  type?: TxDetailType;
};

export type GetTxByAssetPayload = {
  asset: string;
  offset: number;
  limit: number;
  type?: TxDetailType;
};

export type TxDetailData = RemoteData<Error, InlineResponse200>;

export type State = {
  assets: AssetDetailMap;
  assetArray: AssetDetail[];
  pools: string[];
  poolAddressData: Maybe<ThorchainEndpoints>;
  bnbPoolAddress: Maybe<ThorchainEndpoint>;
  poolAddress: Maybe<string>;
  poolData: PoolDataMap;
  stakerPoolData: Maybe<StakerPoolData>;
  stakerPoolDataLoading: boolean;
  stakerPoolDataError: Maybe<Error>;
  runePrice: number;
  basePriceAsset: string; // set base price asset as a RUNE
  priceIndex: PriceDataIndex;
  error: Maybe<Error>;
  poolLoading: boolean;
  txData: TxDetailData;
  apiBasePath: RemoteData<Error, string>;
};
