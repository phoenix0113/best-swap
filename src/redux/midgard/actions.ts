import { AssetSymbol } from '../../types/bepswap';
import {
  GetStakerPoolDataPayload,
  PriceDataIndex,
  AssetDetailMap,
  GetTxByAddressTxIdPayload,
  GetTxByAddressAssetPayload,
  GetTxByAddressPayload,
  GetTxByAssetPayload,
} from './types';
import {
  AssetDetail,
  PoolDetail,
  StakersAssetData,
  ThorchainEndpoints,
  InlineResponse200,
} from '../../types/generated/midgard';

export interface SetAssetsPayload {
  assetDetailIndex: AssetDetailMap;
  assetDetails: AssetDetail[];
}
export const SET_ASSETS = 'SET_ASSETS';
export interface SetAssets {
  type: typeof SET_ASSETS;
  payload: SetAssetsPayload;
}
export const setAssets = (payload: SetAssetsPayload): SetAssets => ({
  type: SET_ASSETS,
  payload,
});

export const GET_POOLS_REQUEST = 'GET_POOLS_REQUEST';
export interface GetPools {
  type: typeof GET_POOLS_REQUEST;
}
export const getPools = (): GetPools => ({
  type: GET_POOLS_REQUEST,
});

export const GET_POOLS_SUCCESS = 'GET_POOLS_SUCCESS';
export interface GetPoolsSuccess {
  type: typeof GET_POOLS_SUCCESS;
  payload: string[];
}
export const getPoolsSuccess = (payload: string[]): GetPoolsSuccess => ({
  type: GET_POOLS_SUCCESS,
  payload,
});
export const GET_POOLS_FAILED = 'GET_POOLS_FAILED';
export interface GetPoolsFailed {
  type: typeof GET_POOLS_FAILED;
  payload: Error;
}
export const getPoolsFailed = (payload: Error): GetPoolsFailed => ({
  type: GET_POOLS_FAILED,
  payload,
});

type GetPoolDataPayload = {
  assets: string[];
  overrideAllPoolData: boolean;
};

export const GET_POOL_DATA_REQUEST = 'GET_POOL_DATA_REQUEST';
export interface GetPoolData {
  type: typeof GET_POOL_DATA_REQUEST;
  payload: GetPoolDataPayload;
}
export const getPoolData = (payload: GetPoolDataPayload): GetPoolData => ({
  type: GET_POOL_DATA_REQUEST,
  payload,
});

type GetPoolDataSuccessPayload = {
  poolDetails: PoolDetail[];
  overrideAllPoolData: boolean;
};
export const GET_POOL_DATA_SUCCESS = 'GET_POOL_DATA_SUCCESS';
export interface GetPoolDataSuccess {
  type: typeof GET_POOL_DATA_SUCCESS;
  payload: GetPoolDataSuccessPayload;
}
export const getPoolDataSuccess = (
  payload: GetPoolDataSuccessPayload,
): GetPoolDataSuccess => ({
  type: GET_POOL_DATA_SUCCESS,
  payload,
});

export const GET_POOL_DATA_FAILED = 'GET_POOL_DATA_FAILED';
export interface GetPoolDataFailed {
  type: typeof GET_POOL_DATA_FAILED;
  payload: Error;
}
export const getPoolDataFailed = (payload: Error): GetPoolDataFailed => ({
  type: GET_POOL_DATA_FAILED,
  payload,
});

export const GET_STAKER_POOL_DATA_REQUEST = 'GET_STAKER_POOL_DATA_REQUEST';
export interface GetStakerPoolData {
  type: typeof GET_STAKER_POOL_DATA_REQUEST;
  payload: GetStakerPoolDataPayload;
}
export const getStakerPoolData = (
  payload: GetStakerPoolDataPayload,
): GetStakerPoolData => ({
  type: GET_STAKER_POOL_DATA_REQUEST,
  payload,
});

export const GET_STAKER_POOL_DATA_SUCCESS = 'GET_STAKER_POOL_DATA_SUCCESS';
export interface GetStakerPoolDataSuccess {
  type: typeof GET_STAKER_POOL_DATA_SUCCESS;
  payload: StakersAssetData[];
}
export const getStakerPoolDataSuccess = (
  payload: StakersAssetData[],
): GetStakerPoolDataSuccess => ({
  type: GET_STAKER_POOL_DATA_SUCCESS,
  payload,
});

export const GET_STAKER_POOL_DATA_FAILED = 'GET_STAKER_POOL_DATA_FAILED';
export interface GetStakerPoolDataFailed {
  type: typeof GET_STAKER_POOL_DATA_FAILED;
  payload: Error;
}
export const getStakerPoolDataFailed = (
  payload: Error,
): GetStakerPoolDataFailed => ({
  type: GET_STAKER_POOL_DATA_FAILED,
  payload,
});

export const GET_POOL_ADDRESSES_REQUEST = 'GET_POOL_ADDRESSES_REQUEST';
export interface GetPoolAddress {
  type: typeof GET_POOL_ADDRESSES_REQUEST;
}
export const getPoolAddress = (): GetPoolAddress => ({
  type: GET_POOL_ADDRESSES_REQUEST,
});

export const GET_POOL_ADDRESSES_SUCCESS = 'GET_POOL_ADDRESSES_SUCCESS';
export interface GetPoolAddressSuccess {
  type: typeof GET_POOL_ADDRESSES_SUCCESS;
  payload: ThorchainEndpoints;
}
export const getPoolAddressSuccess = (
  payload: ThorchainEndpoints,
): GetPoolAddressSuccess => ({
  type: GET_POOL_ADDRESSES_SUCCESS,
  payload,
});

export const GET_POOL_ADDRESSES_FAILED = 'GET_POOL_ADDRESSES_FAILED';
export interface GetPoolAddressFailed {
  type: typeof GET_POOL_ADDRESSES_FAILED;
  payload: Error;
}
export const getPoolAddressFailed = (payload: Error): GetPoolAddressFailed => ({
  type: GET_POOL_ADDRESSES_FAILED,
  payload,
});

export const GET_RUNE_PRICE_REQUEST = 'GET_RUNE_PRICE_REQUEST';
export interface GetRunePrice {
  type: typeof GET_RUNE_PRICE_REQUEST;
}
export const getRunePrice = (): GetRunePrice => ({
  type: GET_RUNE_PRICE_REQUEST,
});
export const SET_BASE_PRICE_ASSET = 'SET_BASE_PRICE_ASSET';
export interface SetBasePriceAsset {
  type: typeof SET_BASE_PRICE_ASSET;
  payload: AssetSymbol;
}
export const setBasePriceAsset = (payload: AssetSymbol): SetBasePriceAsset => ({
  type: SET_BASE_PRICE_ASSET,
  payload,
});

export const SET_PRICE_INDEX = 'SET_PRICE_INDEX';
export interface SetPriceIndex {
  type: typeof SET_PRICE_INDEX;
  payload: PriceDataIndex;
}
export const setPriceIndex = (payload: PriceDataIndex): SetPriceIndex => ({
  type: SET_PRICE_INDEX,
  payload,
});

// get transactions by address
export const GET_TX_BY_ADDRESS = 'GET_TX_BY_ADDRESS';
export interface GetTxByAddress {
  type: typeof GET_TX_BY_ADDRESS;
  payload: GetTxByAddressPayload;
}
export const getTxByAddress = (
  payload: GetTxByAddressPayload,
): GetTxByAddress => ({
  type: GET_TX_BY_ADDRESS,
  payload,
});

export const GET_TX_BY_ADDRESS_SUCCESS = 'GET_TX_BY_ADDRESS_SUCCESS';
export interface GetTxByAddressSuccess {
  type: typeof GET_TX_BY_ADDRESS_SUCCESS;
  payload: InlineResponse200;
}
export const getTxByAddressSuccess = (
  payload: InlineResponse200,
): GetTxByAddressSuccess => ({
  type: GET_TX_BY_ADDRESS_SUCCESS,
  payload,
});

export const GET_TX_BY_ADDRESS_FAILED = 'GET_TX_BY_ADDRESS_FAILED';
export interface GetTxByAddressFailed {
  type: typeof GET_TX_BY_ADDRESS_FAILED;
  payload: Error;
}
export const getTxByAddressFailed = (payload: Error): GetTxByAddressFailed => ({
  type: GET_TX_BY_ADDRESS_FAILED,
  payload,
});

// get transactions by address and txId
export const GET_TX_BY_ADDRESS_TXID = 'GET_TX_BY_ADDRESS_TXID';
export interface GetTxByAddressTxId {
  type: typeof GET_TX_BY_ADDRESS_TXID;
  payload: GetTxByAddressTxIdPayload;
}
export const getTxByAddressTxId = (
  payload: GetTxByAddressTxIdPayload,
): GetTxByAddressTxId => ({
  type: GET_TX_BY_ADDRESS_TXID,
  payload,
});

export const GET_TX_BY_ADDRESS_TXID_SUCCESS = 'GET_TX_BY_ADDRESS_TXID_SUCCESS';
export interface GetTxByAddressTxIdSuccess {
  type: typeof GET_TX_BY_ADDRESS_TXID_SUCCESS;
  payload: InlineResponse200;
}
export const getTxByAddressTxIdSuccess = (
  payload: InlineResponse200,
): GetTxByAddressTxIdSuccess => ({
  type: GET_TX_BY_ADDRESS_TXID_SUCCESS,
  payload,
});

export const GET_TX_BY_ADDRESS_TXID_FAILED = 'GET_TX_BY_ADDRESS_TXID_FAILED';
export interface GetTxByAddressTxIdFailed {
  type: typeof GET_TX_BY_ADDRESS_TXID_FAILED;
  payload: Error;
}
export const getTxByAddressTxIdFailed = (
  payload: Error,
): GetTxByAddressTxIdFailed => ({
  type: GET_TX_BY_ADDRESS_TXID_FAILED,
  payload,
});

// get transactions by address and asset
export const GET_TX_BY_ADDRESS_ASSET = 'GET_TX_BY_ADDRESS_ASSET';
export interface GetTxByAddressAsset {
  type: typeof GET_TX_BY_ADDRESS_ASSET;
  payload: GetTxByAddressAssetPayload;
}
export const getTxByAddressAsset = (
  payload: GetTxByAddressAssetPayload,
): GetTxByAddressAsset => ({
  type: GET_TX_BY_ADDRESS_ASSET,
  payload,
});

export const GET_TX_BY_ADDRESS_ASSET_SUCCESS =
  'GET_TX_BY_ADDRESS_ASSET_SUCCESS';
export interface GetTxByAddressAssetSuccess {
  type: typeof GET_TX_BY_ADDRESS_ASSET_SUCCESS;
  payload: InlineResponse200;
}
export const getTxByAddressAssetSuccess = (
  payload: InlineResponse200,
): GetTxByAddressAssetSuccess => ({
  type: GET_TX_BY_ADDRESS_ASSET_SUCCESS,
  payload,
});

export const GET_TX_BY_ADDRESS_ASSET_FAILED = 'GET_TX_BY_ADDRESS_ASSET_FAILED';
export interface GetTxByAddressAssetFailed {
  type: typeof GET_TX_BY_ADDRESS_ASSET_FAILED;
  payload: Error;
}
export const getTxByAddressAssetFailed = (
  payload: Error,
): GetTxByAddressAssetFailed => ({
  type: GET_TX_BY_ADDRESS_ASSET_FAILED,
  payload,
});

// get transactions by asset
export const GET_TX_BY_ASSET = 'GET_TX_BY_ASSET';
export interface GetTxByAsset {
  type: typeof GET_TX_BY_ASSET;
  payload: GetTxByAssetPayload;
}
export const getTxByAsset = (payload: GetTxByAssetPayload): GetTxByAsset => ({
  type: GET_TX_BY_ASSET,
  payload,
});

export const GET_TX_BY_ASSET_SUCCESS = 'GET_TX_BY_ASSET_SUCCESS';
export interface GetTxByAssetSuccess {
  type: typeof GET_TX_BY_ASSET_SUCCESS;
  payload: InlineResponse200;
}
export const getTxByAssetSuccess = (
  payload: InlineResponse200,
): GetTxByAssetSuccess => ({
  type: GET_TX_BY_ASSET_SUCCESS,
  payload,
});

export const GET_TX_BY_ASSET_FAILED = 'GET_TX_BY_ASSET_FAILED';
export interface GetTxByAssetFailed {
  type: typeof GET_TX_BY_ASSET_FAILED;
  payload: Error;
}
export const getTxByAssetFailed = (payload: Error): GetTxByAssetFailed => ({
  type: GET_TX_BY_ASSET_FAILED,
  payload,
});

export const GET_API_BASEPATH_PENDING = 'GET_API_BASEPATH_PENDING';
export interface GetApiBasePathPending {
  type: typeof GET_API_BASEPATH_PENDING;
}
export const getApiBasePathPending = (): GetApiBasePathPending => ({
  type: GET_API_BASEPATH_PENDING,
});

export const GET_API_BASEPATH_FAILED = 'GET_API_BASEPATH_FAILED';
export interface GetApiBasePathFailed {
  type: typeof GET_API_BASEPATH_FAILED;
  payload: Error;
}
export const getApiBasePathFailed = (payload: Error): GetApiBasePathFailed => ({
  type: GET_API_BASEPATH_FAILED,
  payload,
});

export const GET_API_BASEPATH_SUCCESS = 'GET_API_BASEPATH_SUCCESS';
export interface GetApiBasePathSuccess {
  type: typeof GET_API_BASEPATH_SUCCESS;
  payload: string;
}
export const getApiBasePathSuccess = (
  payload: string,
): GetApiBasePathSuccess => ({
  type: GET_API_BASEPATH_SUCCESS,
  payload,
});

export type MidgardActionTypes =
  | GetPools
  | GetPoolsSuccess
  | GetPoolsFailed
  | GetPoolData
  | GetPoolDataSuccess
  | GetPoolDataFailed
  | GetStakerPoolData
  | GetStakerPoolDataSuccess
  | GetStakerPoolDataFailed
  | GetPoolAddress
  | GetPoolAddressSuccess
  | GetPoolAddressFailed
  | GetRunePrice
  | SetAssets
  | SetBasePriceAsset
  | SetPriceIndex
  | GetTxByAddress
  | GetTxByAddressSuccess
  | GetTxByAddressFailed
  | GetTxByAddressTxId
  | GetTxByAddressTxIdSuccess
  | GetTxByAddressTxIdFailed
  | GetTxByAddressAsset
  | GetTxByAddressAssetSuccess
  | GetTxByAddressAssetFailed
  | GetTxByAsset
  | GetTxByAssetSuccess
  | GetTxByAssetFailed
  | GetApiBasePathPending
  | GetApiBasePathFailed
  | GetApiBasePathSuccess;
