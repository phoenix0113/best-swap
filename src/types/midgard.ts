/**
 * This modules includes types for Midgard,
 * which are not part of any endpoint
 * and therefore not generated by `openapi-generator`
 */

export type Asset = {
  chain?: string;
  symbol?: string;
  ticker?: string;
}
