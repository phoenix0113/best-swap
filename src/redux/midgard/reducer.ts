import { Reducer } from 'redux';
import { initial, success, pending, failure } from '@devexperts/remote-data-ts';

import {
  getBNBPoolAddress,
  getPoolAddress,
  getPriceIndex,
  getAssetSymbolFromPayload,
} from './utils';
import { getBasePriceAsset } from '../../helpers/webStorageHelper';
import { State, PoolDataMap, StakerPoolData } from './types';
import {
  SET_BASE_PRICE_ASSET,
  SET_PRICE_INDEX,
  GET_RUNE_PRICE_REQUEST,
  SET_ASSETS,
  GET_POOLS_REQUEST,
  GET_POOLS_SUCCESS,
  GET_POOLS_FAILED,
  GET_POOL_DATA_SUCCESS,
  GET_POOL_DATA_FAILED,
  GET_POOL_DATA_REQUEST,
  MidgardActionTypes,
  GET_STAKER_POOL_DATA_SUCCESS,
  GET_STAKER_POOL_DATA_REQUEST,
  GET_STAKER_POOL_DATA_FAILED,
  GET_POOL_ADDRESSES_SUCCESS,
  GET_POOL_ADDRESSES_REQUEST,
  GET_POOL_ADDRESSES_FAILED,
  GET_TX_BY_ADDRESS,
  GET_TX_BY_ADDRESS_SUCCESS,
  GET_TX_BY_ADDRESS_FAILED,
  GET_TX_BY_ADDRESS_TXID,
  GET_TX_BY_ADDRESS_TXID_SUCCESS,
  GET_TX_BY_ADDRESS_TXID_FAILED,
  GET_TX_BY_ADDRESS_ASSET,
  GET_TX_BY_ADDRESS_ASSET_SUCCESS,
  GET_TX_BY_ADDRESS_ASSET_FAILED,
  GET_TX_BY_ASSET,
  GET_TX_BY_ASSET_SUCCESS,
  GET_TX_BY_ASSET_FAILED,
} from './actions';
import { Nothing } from '../../types/bepswap';
import { PoolDetail, StakersAssetData } from '../../types/generated/midgard';

const basePriceAsset = getBasePriceAsset() || 'RUNE';

const initState: State = {
  assets: {},
  assetArray: [],
  pools: [],
  poolAddressData: Nothing,
  bnbPoolAddress: Nothing,
  poolAddress: Nothing,
  poolData: {},
  stakerPoolData: {},
  runePrice: 0,
  basePriceAsset, // set base price asset as a RUNE
  priceIndex: {
    RUNE: 1,
  },
  error: null,
  poolLoading: false,
  stakerPoolDataLoading: false,
  txData: initial,
};

const reducer: Reducer<State, MidgardActionTypes> = (
  state = initState,
  action,
) => {
  switch (action.type) {
    case SET_BASE_PRICE_ASSET: {
      const { payload } = action;
      return {
        ...state,
        basePriceAsset: payload,
        priceIndex: getPriceIndex(state.assetArray, payload),
      };
    }
    case SET_PRICE_INDEX:
      return {
        ...state,
        priceIndex: action.payload,
      };
    case GET_RUNE_PRICE_REQUEST:
      return {
        ...state,
        runePrice: 0,
        error: Nothing,
      };
    case SET_ASSETS: {
      const { payload } = action;
      return {
        ...state,
        assets: payload.assetDetailIndex,
        assetArray: payload.assetDetails,
      };
    }
    case GET_POOLS_REQUEST:
      return {
        ...state,
        poolLoading: true,
        error: Nothing,
      };
    case GET_POOLS_SUCCESS:
      return {
        ...state,
        poolLoading: false,
        pools: action.payload,
      };
    case GET_POOLS_FAILED:
      return {
        ...state,
        poolLoading: false,
        error: action.payload,
      };
    case GET_POOL_DATA_REQUEST:
      return {
        ...state,
        poolLoading: true,
        error: Nothing,
      };
    case GET_POOL_DATA_SUCCESS: {
      const { payload } = action;
      const { poolDetails = [], overrideAllPoolData = true } = payload;
      // Transform `PoolDetail[]` into `PoolDataMap` before storing data into state
      const newPoolData = poolDetails.reduce(
        (acc: PoolDataMap, poolDetail: PoolDetail) => {
          const symbol = getAssetSymbolFromPayload(poolDetail);
          return symbol
            ? {
                ...acc,
                [symbol]: poolDetail,
              }
            : acc;
        },
        {} as PoolDataMap,
      );
      // Check whether to override all state.poolData or just with latest result
      const poolData = overrideAllPoolData
        ? newPoolData
        : {
            ...state.poolData,
            newPoolData,
          };
      return {
        ...state,
        poolData,
        poolLoading: false,
      };
    }
    case GET_POOL_DATA_FAILED:
      return {
        ...state,
        poolLoading: false,
        error: action.payload,
      };
    case GET_STAKER_POOL_DATA_REQUEST:
      return {
        ...state,
        stakerPoolDataLoading: true,
        error: Nothing,
      };
    case GET_STAKER_POOL_DATA_SUCCESS: {
      const { payload } = action;
      // Transform `StakersAssetData[]` into `StakerPoolData`
      // before storing data into state
      const newStakerPoolData = payload.reduce(
        (acc: StakerPoolData, data: StakersAssetData) => {
          const symbol = getAssetSymbolFromPayload(data);
          return symbol ? { ...acc, [symbol]: data } : acc;
        },
        {} as StakerPoolData,
      );

      return {
        ...state,
        stakerPoolData: { ...state.stakerPoolData, ...newStakerPoolData },
        stakerPoolDataLoading: false,
      };
    }
    case GET_STAKER_POOL_DATA_FAILED:
      return {
        ...state,
        stakerPoolData: {},
        stakerPoolDataLoading: false,
        error: action.payload,
      };
    case GET_POOL_ADDRESSES_REQUEST:
      return {
        ...state,
        error: Nothing,
      };
    case GET_POOL_ADDRESSES_SUCCESS: {
      const { payload } = action;
      return {
        ...state,
        poolAddressData: payload,
        bnbPoolAddress: getBNBPoolAddress(payload),
        poolAddress: getPoolAddress(payload),
      };
    }
    case GET_POOL_ADDRESSES_FAILED:
      return {
        ...state,
        poolAddressData: Nothing,
        bnbPoolAddress: {},
        poolAddress: Nothing,
        error: action.payload,
      };
    case GET_TX_BY_ADDRESS:
      return {
        ...state,
        txData: pending,
      };
    case GET_TX_BY_ADDRESS_SUCCESS:
      return {
        ...state,
        txData: success(action.payload),
      };
    case GET_TX_BY_ADDRESS_FAILED:
      return {
        ...state,
        txData: failure(action.payload),
      };
    case GET_TX_BY_ADDRESS_ASSET:
      return {
        ...state,
        txData: pending,
      };
    case GET_TX_BY_ADDRESS_ASSET_SUCCESS:
      return {
        ...state,
        txData: success(action.payload),
      };
    case GET_TX_BY_ADDRESS_ASSET_FAILED:
      return {
        ...state,
        txData: failure(action.payload),
      };

    case GET_TX_BY_ADDRESS_TXID:
      return {
        ...state,
        txData: pending,
      };
    case GET_TX_BY_ADDRESS_TXID_SUCCESS:
      return {
        ...state,
        txData: success(action.payload),
      };
    case GET_TX_BY_ADDRESS_TXID_FAILED:
      return {
        ...state,
        txData: failure(action.payload),
      };
    case GET_TX_BY_ASSET:
      return {
        ...state,
        txData: pending,
      };
    case GET_TX_BY_ASSET_SUCCESS:
      return {
        ...state,
        txData: success(action.payload),
      };
    case GET_TX_BY_ASSET_FAILED:
      return {
        ...state,
        txData: failure(action.payload),
      };
    default:
      return state;
  }
};
export default reducer;
