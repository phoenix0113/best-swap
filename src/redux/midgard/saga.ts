import { all, takeEvery, put, fork, call } from 'redux-saga/effects';
import { AxiosResponse } from 'axios';
import { isEmpty as _isEmpty } from 'lodash';
import * as actions from './actions';
import { MIDGARD_API_URL } from '../../helpers/apiHelper';

import {
  saveBasePriceAsset,
  getBasePriceAsset,
} from '../../helpers/webStorageHelper';
import { getAssetDetailIndex, getPriceIndex } from './utils';
import {
  DefaultApi,
  AssetDetail,
  PoolDetail,
  StakersAssetData,
  ThorchainEndpoints,
  TxDetails,
} from '../../types/generated/midgard';

const midgardApi = new DefaultApi({ basePath: MIDGARD_API_URL });

export function* getPools() {
  yield takeEvery(actions.GET_POOLS_REQUEST, function*() {
    try {
      const { data }: AxiosResponse<string[]> = yield call({
        context: midgardApi,
        fn: midgardApi.getPools,
      });

      if (data && !_isEmpty(data)) {
        const { data: assetDetails }: AxiosResponse<AssetDetail[]> = yield call(
          {
            context: midgardApi,
            fn: midgardApi.getAssetInfo,
          },
          data.join(),
        );

        const assetDetailIndex = getAssetDetailIndex(assetDetails);
        const baseTokenTicker = getBasePriceAsset() || 'RUNE';
        const priceIndex = getPriceIndex(assetDetails, baseTokenTicker);
        const assetsPayload: actions.SetAssetsPayload = {
          assetDetails,
          assetDetailIndex,
        };

        yield put(actions.setAssets(assetsPayload));
        yield put(actions.setPriceIndex(priceIndex));

        yield put(
          actions.getPoolData({ assets: data, overrideAllPoolData: true }),
        );

        yield put(actions.getPoolsSuccess(data));
      } else {
        const error = new Error('No pools available');
        yield put(actions.getPoolsFailed(error));
      }
    } catch (error) {
      yield put(actions.getPoolsFailed(error));
    }
  });
}

export function* getPoolData() {
  yield takeEvery(actions.GET_POOL_DATA_REQUEST, function*({
    payload,
  }: actions.GetPoolData) {
    const { assets, overrideAllPoolData } = payload;
    try {
      const { data }: AxiosResponse<PoolDetail[]> = yield call(
        { context: midgardApi, fn: midgardApi.getPoolsData },
        assets.join(),
      );
      yield put(
        actions.getPoolDataSuccess({ poolDetails: data, overrideAllPoolData }),
      );
    } catch (error) {
      yield put(actions.getPoolDataFailed(error));
    }
  });
}

export function* getStakerPoolData() {
  yield takeEvery(actions.GET_STAKER_POOL_DATA_REQUEST, function*({
    payload,
  }: actions.GetStakerPoolData) {
    const { address, asset } = payload;

    // TODO (Chris): currently hardcode the Chain as BNB
    const assetId = `BNB.${asset}`;

    try {
      const { data }: AxiosResponse<StakersAssetData> = yield call(
        { context: midgardApi, fn: midgardApi.getStakersAddressAndAssetData },
        address,
        assetId,
      );

      yield put(actions.getStakerPoolDataSuccess(data));
    } catch (error) {
      yield put(actions.getStakerPoolDataFailed(error));
    }
  });
}

export function* getPoolAddress() {
  yield takeEvery(actions.GET_POOL_ADDRESSES_REQUEST, function*() {
    try {
      const { data }: AxiosResponse<ThorchainEndpoints> = yield call({
        context: midgardApi,
        fn: midgardApi.getThorchainProxiedEndpoints,
      });

      yield put(actions.getPoolAddressSuccess(data));
    } catch (error) {
      yield put(actions.getPoolAddressFailed(error));
    }
  });
}

export function* getTxByAddress() {
  yield takeEvery(actions.GET_TX_BY_ADDRESS, function*({
    payload,
  }: actions.GetTxByAddress) {
    try {
      const { data }: AxiosResponse<TxDetails[]> = yield call(
        {
          context: midgardApi,
          fn: midgardApi.getTxDetails,
        },
        payload,
      );

      yield put(actions.getTxByAddressSuccess(data));
    } catch (error) {
      yield put(actions.getTxByAddressFailed(error));
    }
  });
}

export function* getTxByAddressTxId() {
  yield takeEvery(actions.GET_TX_BY_ADDRESS_TXID, function*({
    payload,
  }: actions.GetTxByAddressTxId) {
    try {
      const { address, txId } = payload;
      const { data }: AxiosResponse<TxDetails[]> = yield call(
        {
          context: midgardApi,
          fn: midgardApi.getTxDetailsByAddressTxId,
        },
        address,
        txId,
      );

      yield put(actions.getTxByAddressTxIdSuccess(data));
    } catch (error) {
      yield put(actions.getTxByAddressTxIdFailed(error));
    }
  });
}

export function* getTxByAddressAsset() {
  yield takeEvery(actions.GET_TX_BY_ADDRESS_ASSET, function*({
    payload,
  }: actions.GetTxByAddressAsset) {
    try {
      const { address, asset } = payload;
      const { data }: AxiosResponse<TxDetails[]> = yield call(
        {
          context: midgardApi,
          fn: midgardApi.getTxDetailsByAddressAsset,
        },
        address,
        asset,
      );

      yield put(actions.getTxByAddressAssetSuccess(data));
    } catch (error) {
      yield put(actions.getTxByAddressAssetFailed(error));
    }
  });
}

export function* getTxByAsset() {
  yield takeEvery(actions.GET_TX_BY_ASSET, function*({
    payload,
  }: actions.GetTxByAsset) {
    try {
      const { data }: AxiosResponse<TxDetails[]> = yield call(
        {
          context: midgardApi,
          fn: midgardApi.getTxDetailsByAsset,
        },
        payload,
      );

      yield put(actions.getTxByAssetSuccess(data));
    } catch (error) {
      yield put(actions.getTxByAssetFailed(error));
    }
  });
}

export function* setBasePriceAsset() {
  yield takeEvery(actions.SET_BASE_PRICE_ASSET, function*({
    payload,
  }: actions.SetBasePriceAsset) {
    yield call(saveBasePriceAsset, payload);
  });
}

export default function* rootSaga() {
  yield all([
    fork(getPools),
    fork(getPoolData),
    fork(getStakerPoolData),
    fork(getPoolAddress),
    fork(setBasePriceAsset),
    fork(getTxByAddress),
    fork(getTxByAddressTxId),
    fork(getTxByAddressAsset),
    fork(getTxByAsset),
  ]);
}
