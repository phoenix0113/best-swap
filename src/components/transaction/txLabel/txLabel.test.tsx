import React from 'react';
import { mount } from 'enzyme';
import { create } from 'react-test-renderer';

import TxLabel from './txLabel';
import { TxDetailsTypeEnum } from '../../../types/generated/midgard';

describe('FilterDropdown', () => {
  const props = {
    type: TxDetailsTypeEnum.Swap,
  };
  const component = mount(<TxLabel {...props} />);

  it('Matches the snapshot', () => {
    const snapshot = create(<TxLabel {...props} />);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('render wrapper div', () => {
    expect(component.find('div.txLabel-wrapper')).toHaveLength(1);
  });
  it('has label for tx type', () => {
    expect(component.find('p')).toHaveLength(1);
  });
  it('has label - swap for tx type sell', () => {
    const sellTypeComponent = mount(<TxLabel type={TxDetailsTypeEnum.Swap} />);

    expect(sellTypeComponent.find('p').text()).toBe('swap');
  });

  it('has label - stake for tx type stake', () => {
    const stakeTypeComponent = mount(<TxLabel type={TxDetailsTypeEnum.Stake} />);

    expect(stakeTypeComponent.find('p').text()).toBe('stake');
  });
  it('has label - withdraw for tx type withdraw', () => {
    const withdrawTypeComponent = mount(<TxLabel type={TxDetailsTypeEnum.Unstake} />);

    expect(withdrawTypeComponent.find('p').text()).toBe('withdraw');
  });
  it('has icon', () => {
    expect(component.find('.tx-label-icon')).toHaveLength(1);
    expect(
      component
        .find('.tx-label-icon')
        .children()
        .find('svg'),
    ).toHaveLength(1);
  });
});
