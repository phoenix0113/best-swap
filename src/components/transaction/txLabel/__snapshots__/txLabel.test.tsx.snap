// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterDropdown Matches the snapshot 1`] = `
<div
  className="sc-AxjAm dzgnjC txLabel-wrapper"
>
  <p>
    swap
  </p>
  <div
    className="tx-label-icon"
  >
    <svg>
      tx-swap.svg
    </svg>
  </div>
</div>
`;
